### PR TITLE
Support for Azure CNI Cilium

### LINKED ISSUE #397: Support for Azure CNI - Cilium
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Azure CNI powered by Cilium in now generally available. Cilium is a comprehensive networking solution that combines the control plane of Azure CNI with the dataplane capabilities of Cilium.

Resources:
- https://azure.microsoft.com/en-us/updates/azure-cni-powered-by-cilium
- https://learn.microsoft.com/en-us/azure/aks/azure-cni-powered-by-cilium

Status:
- Implementation in the Terraform provider is available.
  https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#ebpf_data_plane

Validation:
- When ebpf_data_plane is set to cilium, the network_plugin field can only be set to azure.
- When ebpf_data_plane is set to cilium, one of either network_plugin_mode = "Overlay" or pod_subnet_id must be specified.

### New or Affected Resource(s)/Data Source(s)

azurerm_kubernetes_cluster

### Potential Terraform Configuration

```hcl
variable "ebpf_data_plane" {
  type        = string
  default     = null
  description = "(Optional) Specifies the eBPF data plane used for building the Kubernetes network."
}

resource "azurerm_kubernetes_cluster" "main" {
  network_profile {
    [..CUT..]
    ebpf_data_plane = var.ebpf_data_plane
    [..CUT..]
  }
  lifecycle {
    precondition {
      condition     = !(var.ebpf_data_plane == "cilium" && var.network_plugin != "azure")
      error_message = "When ebpf_data_plane is set to cilium, the network_plugin field can only be set to azure."
    }
    precondition {
      condition     = !(var.ebpf_data_plane == "cilium") || ( var.network_plugin_mode == "Overlay" && var.pod_subnet_id == null || var.network_plugin_mode == null && var.pod_subnet_id != null )
      error_message = "When ebpf_data_plane is set to cilium, one of either network_plugin_mode = `Overlay` or pod_subnet_id must be specified."
    }
  }
}
```

### References

_No response_

### PR DESCRIPTION
**Pull request by JitseHijlkema**
## Describe your changes
Support for Azure CNI - Cilium

Azure CNI powered by Cilium in now generally available. 
Cilium is a comprehensive networking solution that combines the control plane of Azure CNI with the dataplane capabilities of Cilium.

## Issue number

Closes #397

## Checklist before requesting a review
- [x] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [x] I have executed pre-commit on my machine
- [x] I have passed pr-check on my machine

### CODE REVIEW

**Comment by JitseHijlkema** [comment]
@microsoft-github-policy-service agree

**Comment by zioproto** [comment]
Please run the tests locally as described in the Readme:
https://github.com/Azure/terraform-azurerm-aks#pre-commit--pr-check--test

File: main.tf [hunk 1]
```diff
@@ -483,6 +484,14 @@ resource "azurerm_kubernetes_cluster" "main" {
       condition     = var.network_plugin_mode != "Overlay" || var.network_plugin == "azure"
       error_message = "When network_plugin_mode is set to Overlay, the network_plugin field can only be set to azure."
     }
+    precondition {
+      condition     = !(var.ebpf_data_plane == "cilium" && var.network_plugin != "azure")
+      error_message = "When ebpf_data_plane is set to cilium, the network_plugin field can only be set to azure."
+    }
+    precondition {
+      condition     = !(var.ebpf_data_plane == "cilium") || (var.network_plugin_mode == "Overlay" && var.pod_subnet_id == null || var.network_plugin_mode == null && var.pod_subnet_id != null)
```
Comment (lonegunmanb) [inline main.tf:492]: What if `var.ebpf_data_plan == "cilium" && var.network_plugin_mode == "Overlay" && `var.pod_subnet_id != null`? Is that a valid combination?

Comment (lonegunmanb) [inline main.tf:492]: According to the message, I think we can simplify the condition to:

```hcl
condition     = (var.ebpf_data_plane != "cilium") || (var.network_plugin_mode == "Overlay" || var.pod_subnet_id != null)
```

File: variables.tf [hunk 2]
```diff
@@ -703,6 +703,12 @@ variable "network_policy" {
   description = " (Optional) Sets up network policy to be used with Azure CNI. Network policy allows us to control the traffic flow between pods. Currently supported values are calico and azure. Changing this forces a new resource to be created."
 }
 
+variable "ebpf_data_plane" {
+  type        = string
+  default     = null
+  description = "(Optional) Specifies the eBPF data plane used for building the Kubernetes network."
```
Comment (lonegunmanb) [inline variables.tf:709]: Description could be 

```text
(Optional) Specifies the eBPF data plane used for building the Kubernetes network. Possible value is `cilium`. Changing this forces a new resource to be created.
```

File: main.tf [hunk 3]
```diff
@@ -483,6 +484,14 @@ resource "azurerm_kubernetes_cluster" "main" {
       condition     = var.network_plugin_mode != "Overlay" || var.network_plugin == "azure"
       error_message = "When network_plugin_mode is set to Overlay, the network_plugin field can only be set to azure."
     }
+    precondition {
+      condition     = !(var.ebpf_data_plane == "cilium" && var.network_plugin != "azure")
```
Comment (lonegunmanb) [inline main.tf:488]: Could we simplify the condition to `var.ebpf_data_plan != "cilium" || var.network_plugin == "azure"`?

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @JitseHijlkema for opening this pr! Some comments.

File: variables.tf (hunk 2 above)
Comment (JitseHijlkema) [reply variables.tf:709]: Agree, modified :)

File: main.tf (hunk 1 above)
Comment (JitseHijlkema) [reply main.tf:492]: Slightly modified, to still comply to 'one of either'

File: main.tf [hunk 4]
```diff
@@ -483,6 +484,14 @@ resource "azurerm_kubernetes_cluster" "main" {
       condition     = var.network_plugin_mode != "Overlay" || var.network_plugin == "azure"
       error_message = "When network_plugin_mode is set to Overlay, the network_plugin field can only be set to azure."
     }
+    precondition {
+      condition     = var.ebpf_data_plane != "cilium" || var.network_plugin == "azure"
+      error_message = "When ebpf_data_plane is set to cilium, the network_plugin field can only be set to azure."
+    }
+    precondition {
+      condition     = (var.ebpf_data_plane != "cilium") || (var.network_plugin_mode == "Overlay" && var.pod_subnet_id == null || var.pod_subnet_id != null && var.network_plugin_mode == null)
```
Comment (zioproto) [inline main.tf:492]: You can have `var.network_plugin_mode == "Overlay"` and specify a value for example `var.pod_subnet_id == 10.100.0.0/16`. This is a valid configuration with Cilium but the current precondition would not allow it.

Can you please double check ?

Would it work this one? :
```
(var.ebpf_data_plane != "cilium") || var.network_plugin_mode == "Overlay" || var.pod_subnet_id != null
```
Also here you are just checking `var.pod_subnet_id` for the default node pool, but the additional nodepools are not part for the precondition

Comment (JitseHijlkema) [reply main.tf]: Yes, I updated the precondition.

**Review by lonegunmanb — APPROVED** [review]
Thanks @JitseHijlkema for the update, LGTM!

**Pull request merged**